File: Shouty/Shouty.Specs/StepDefinitions/StepDefinitions.cs
using System;
using System.Collections.Generic;
using System.Linq;
using TechTalk.SpecFlow;
using Xunit;

namespace Shouty.Specs.StepDefinitions
{
    [Binding]
    public class StepDefinitions
    {
        private const int DEFAULT_RANGE = 100;

        private string messageFromSean;
        private Network network = new Network(DEFAULT_RANGE);
        private Dictionary<string, Person> people;

        [BeforeScenario]
        public void CreateNetwork()
        {
            people = new Dictionary<string, Person>();
        }

        [Given("the range is {int}")]
        public void GivenTheRangeIs(int range)
        {
            network = new Network(range);
        }

        [Given("a person named {word}")]
        public void GivenAPersonNamed(string name)
        {
            people.Add(name, new Person(network, 0));
        }

        [Given("a person named {word} is located at {int}")]
        public void GivenAPersonNamedIsLocatedAt(string name, int location)
        {
            people.Add(name, new Person(network, location));
        }

        [Given("people are located at")]
        public void GivenPeopleAreLocatedAt(Table personTable)
        {
            foreach (var row in personTable.Rows)
            {
                people.Add(row["name"], new Person(network, int.Parse(row["location"])));
            }
        }

        [When("Sean shouts")]
        public void WhenSeanShouts()
        {
            people["Sean"].Shout("Hello, world");
        }

        [When("Sean shouts {string}")]
        public void WhenSeanShoutsAMessage(string message)
        {
            people["Sean"].Shout(message);
            messageFromSean = message;
        }

        [Then("Lucy should hear Sean's message")]
        public void ThenLucyShouldHearSeansMessage()
        {
            Assert.Contains(messageFromSean, people["Lucy"].GetMessagesHeard());
        }

        [Then("Lucy should hear a shout")]
        public void ThenLucyShouldHearAShout()
        {
            Assert.Equal(1, people["Lucy"].GetMessagesHeard().Count);
        }

        [Then("Larry should not hear a shout")]
        public void ThenLarryShouldNotHearAShout()
        {
            Assert.Equal(0, people["Larry"].GetMessagesHeard().Count);
        }
    }
}

File: content/04/code/java/12-replace-hardcoded-lucy-instance/shouty/src/test/java/shouty/StepDefinitions.java
package shouty;

import io.cucumber.java.Before;
import io.cucumber.java.en.Given;
import io.cucumber.java.en.Then;
import io.cucumber.java.en.When;

import java.util.HashMap;

import static java.util.Arrays.asList;
import static org.junit.Assert.assertEquals;

public class StepDefinitions {

    private Person sean;
    private Person lucy;
    private String messageFromSean;
    private Network network;
    private HashMap<String, Person> people;

    @Before
    public void createNetwork() {
        network = new Network();
        people = new HashMap<>();
    }

    @Given("a person named Lucy")
    public void a_person_named_Lucy() {
        people.put("Lucy", new Person(network));
    }

    @Given("a person named Sean")
    public void a_person_named_Sean() {
      sean = new Person(network);
    }

    @When("Sean shouts {string}")
    public void sean_shouts(String message) throws Throwable {
        sean.shout(message);
        messageFromSean = message;
    }

    @Then("Lucy should hear Sean's message")
    public void lucy_hears_Sean_s_message() throws Throwable {
        assertEquals(asList(messageFromSean), people.get("Lucy").getMessagesHeard());
    }
}
